pkg/runx/impl/registry/artifact.go

```diff
@@ -7,25 +7,33 @@ import (
 	"go.jetpack.io/pkg/runx/impl/types"
 )
 
-func findArtifactForPlatform(artifacts []types.ArtifactMetadata, platform types.Platform) *types.ArtifactMetadata {
-	var artifactForPlatform *types.ArtifactMetadata
+func findArtifactForPlatform(artifacts []types.ArtifactMetadata, platform types.Platform) (types.ArtifactMetadata, error) {
+	platformCompatible := false
+	knownArchive := false
+
 	for _, artifact := range artifacts {
-		if isArtifactForPlatform(artifact, platform) {
-			artifactForPlatform = &artifact
-			if isKnownArchive(artifact.Name) {
-				// We only consider known archives because sometimes releases contain multiple files
-				// for the same platform. Some times those files are alternative installation methods
-				// like `.dmg`, `.msi`, or `.deb`, and sometimes they are metadata files like `.sha256`
-				// or a `.sig` file. We don't want to install those.
-				return &artifact
-			}
+		if isArtifactForPlatform(artifact.Name, platform) {
+			platformCompatible = true
+		}
+		if isKnownArchive(artifact.Name) {
+			// We only consider known archives because sometimes releases contain multiple files
+			// for the same platform. Some times those files are alternative installation methods
+			// like `.dmg`, `.msi`, or `.deb`, and sometimes they are metadata files like `.sha256`
+			// or a `.sig` file. We don't want to install those.
+			knownArchive = true
 		}
+
+		if platformCompatible && knownArchive {
+			return artifact, nil
+		}
```

Comment:
Because `platformCompatible` and `knownArchive` get set to true and never get reset it's possible for two different artifacts to make this statement true when combined even if each individual one doesn't actually satisfy. (in practice this is unlikely, but not impossible). You can test this by adding 2 different artifacts, one only satisfies `isArtifactForPlatform` and the other only satisfies `isKnownArchive`.

Here's a small variation on your logic that maintains your goal of splitting no-platform and no-archive errors:

```golang
func findArtifactForPlatform(artifacts []types.ArtifactMetadata, platform types.Platform) (types.ArtifactMetadata, error) {
	foundPlatform := false
	for _, artifact := range artifacts {
		if isArtifactForPlatform(artifact.Name, platform) {
			foundPlatform = true
			if isKnownArchive(artifact.Name) {
				// We only consider known archives because sometimes releases contain multiple files
				// for the same platform. Some times those files are alternative installation methods
				// like `.dmg`, `.msi`, or `.deb`, and sometimes they are metadata files like `.sha256`
				// or a `.sig` file. We don't want to install those.
				return artifact, nil
			}
		}
	}
	if !foundPlatform {
		return types.ArtifactMetadata{}, types.ErrPlatformNotSupported
	}
	return types.ArtifactMetadata{}, types.ErrNoKnownArchive
}
```

Reply:
@mikeland73 oh i see your point. you're right. thank you for the code!